Set temporary single CPU affinity before cgroup cpuset transition.

cclerget opened the pull request:
This handles a corner case when joining a container having all the processes running exclusively on isolated CPU cores to force the kernel to schedule runc process on the first CPU core within the cgroups cpuset.

The introduction of the kernel commit
46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76 has affected this deterministic scheduling behavior by distributing tasks across CPU cores within the cgroups cpuset. Some intensive real-time application are relying on this deterministic behavior and use the first CPU core to run a slow thread while other CPU cores are fully used by real-time threads with SCHED_FIFO policy. Such applications prevents runc process from joining a container when the runc process is randomly scheduled on a CPU core owned by a real-time thread.

* Fixes #3922

kolyshkin commented on libcontainer/container_linux.go:
```diff
@@ -2261,3 +2290,61 @@ func requiresRootOrMappingTool(c *configs.Config) bool {
+// - when cpuset cores are all isolated cores: return first CPU of the cpuset
+// - when cpuset cores are mixed between housekeeping/isolated cores: no eligible CPU
+func getEligibleCPU(cpusetList, isolatedList string) (int, bool, error) {
+	if isolatedList == "" {
+		return 0, false, nil
+	} else if cpusetList == "" {
```
you can combine these two into
```go
if isolatedList == "" || cpusetList == "" {
```

kolyshkin commented on libcontainer/container_linux.go:
```diff
@@ -2261,3 +2290,61 @@ func requiresRootOrMappingTool(c *configs.Config) bool {
+		return 0, false, nil
+	}
+
+	// The target container has a cgroup cpuset, get the bit range
```
nit: missing period

kolyshkin commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -1250,6 +1260,29 @@ void nsexec(void)
+			 * cgroups cpuset. Some intensive real-time application are relying on this
+			 * deterministic behavior and use the first CPU core to run a slow thread
+			 * while other CPU cores are fully used by real-time threads with SCHED_FIFO
+			 * policy. Such applications prevents runc process from joining a container
```
nit: `s/prevents/prevent/`

kolyshkin commented on libcontainer/container_linux_test.go:
```diff
@@ -390,3 +390,87 @@ func TestGetContainerStateAfterUpdate(t *testing.T) {
+		t.Run(tt.name, func(t *testing.T) {
+			affinityCore, eligible, err := getEligibleCPU(tt.cpuset, tt.isolset)
+			if err != nil && !tt.expectedErr {
+				t.Fatalf("unexpected error for %q: %s", tt.name, err)
+			} else if err == nil && tt.expectedErr {
+				t.Fatalf("unexpected success for %q", tt.name)
+			} else if tt.expectedEligible && !eligible {
+				t.Fatalf("was expecting eligible core but no eligible core returned for %q", tt.name)
```
nit: no need to add tt.name to error text, as it added to the test name by `t.Run`.

kolyshkin commented:
Thanks for working on this. I changed it to draft until all the issues with the code and test cases are fixed, and left some minor comments.

cclerget commented:
Thanks @kolyshkin , it should be ready for review now

lifubang commented:
@cclerget Please rebase

cclerget commented:
@lifubang Done

lifubang commented on libcontainer/container_linux.go:
```diff
@@ -2280,3 +2309,57 @@ func requiresRootOrMappingTool(c *configs.Config) bool {
+// - when there is not cpuset cores: no eligible CPU
+// - when there is not isolated cores: no eligible CPU
+// - when cpuset cores are all isolated cores: return first CPU of the cpuset
+// - when cpuset cores are mixed between housekeeping/isolated cores: no eligible CPU.
```
@cclerget Is this rule too strict? I think we can calculate an eligible CPU. What was your idea when making this rule? Is it about algorithms’ complexity? Or something else I didn't notice?

cclerget replied:
@lifubang Yes, originally in kubernetes context, I translated this to a non-deterministic CPU allocation and I wasn't expecting such setup to be used. But you're right, the first housekeeping CPU could be used here to handle specific CPU allocations no matter the context

lifubang commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -1053,6 +1063,29 @@ void nsexec(void)
+				cpu_set_t set;
+				CPU_ZERO(&set);
+				CPU_SET(config.cpu_affinity, &set);
+				if (sched_setaffinity(0, sizeof(set), &set) == -1)
```
@cclerget I have tested your solution in my host. There is a big problem, the cpu affinity has the higher priority then `cpuset.cpus`. All tasks in the container will be scheduled to this tempporary single CPU.
I think you should remove the cpu affinity before the container started.

cclerget replied:
@lifubang When the process is joining an existing container, the CPU affinity mask is reset to the cgroup cpuset, `taskset -c -p $$` returns the cpu affinity mask corresponding to cpuset.

What you've observed is expected, as the scheduler do not schedule tasks pinned to isolated CPUs, that's why this [patch](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76) is problematic.
Since the scheduler doesn't manage tasks on isolated CPUs, everything is executed on the temporary CPU, my changes just restore the kernel behavior prior to kernel patch above.
You can test without my changes, you will also observe that all processes runs on the same isolated CPU, the difference is that the isolated CPU will randomly change each time you are entering in the container, but subsequent processes still runs on the same isolated CPU.

Removing the cpu affinity before executing container process would defeat the purpose of this patch, as the kernel could pin it randomly on a "locked" isolated core, this is the last case from [patch](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76) description:

```
The cases that this mainly affects are:
 - modifying cpuset.cpus
 - when tasks join a cpuset
 - when modifying a task's affinity via sched_setaffinity(2)
```

lifubang replied:
> What you've observed is expected,

I think it is not expected.
For example:
```bash
root@acmcoder:/home/acmcoder# cat /sys/devices/system/cpu/isolated
0-1,3
```
ContainerA: cpuset.cpus: "1,3"
When the ContainerA started, all the tasks in this container will be scheduled to `cpu1`, no tasks in `cpu3`.

lifubang replied:
> You can test without my changes, you will also observe that all processes runs on the same isolated CPU,

No, I have tested it. With the same configuration as above, both `cpu1` and `cpu3` have  the same level load.

cclerget replied:
> ContainerA: cpuset.cpus: "1,3"
When the ContainerA started, all the tasks in this container will be scheduled to cpu1, no tasks in cpu3.

Just to be sure, the patch is affecting `runc exec` not `runc run` (`initSetns` only https://github.com/opencontainers/runc/pull/3923/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2251)

Let me clarify and describe what I did for testing, I have a VM with 8 CPUs with `isolcpus=0-3`.

I'm using an OCI config containing, the container just call sleep for some time :

```
	"linux": {
		"cgroupsPath": "/repro",
		"resources": {
			"cpu": {
				"cpus": "0-3",
				"mems": "0"
			},
			"devices": [
				{
					"allow": false,
					"access": "rwm"
				}
			]
		},
		"namespaces": [
			{
				"type": "mount"
			}
		]
	}
```

In a first terminal I'm starting the container with:

```
sudo runc run --pid-file /tmp/repro.pid --bundle /tmp/bundle repro
```

Then in a second terminal I'm spawing another process to test the patch :

```
sudo runc exec repro sh -c "for i \$(seq 1 100); do cat /proc/self/stat|awk '{print $39}'; done"
```

I did that with and without my runc changes and with `isolcpus=0-3` for both, all iterations will show the exact same CPU between 0-3 for the unmodified runc (this is expected, this is the kernel [patch](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76) in action):

```
sudo runc exec repro sh -c "for i \$(seq 1 100); do cat /proc/self/stat|cut -d ' ' -f 39; done"
2
2
2
2
2
2
....
```

While modified runc will show 0 for all (expected result), my patch spawn the entering process on the first CPU of the `cpuset.cpus`.


```
sudo runc exec repro sh -c "for i \$(seq 1 100); do cat /proc/self/stat|cut -d ' ' -f 39; done"
2
2
2
2
2
2
....
```
Now if you run the same without `isolcpus` set in the VM, you will notice that all iterations are showing different CPU for each iterations.

lifubang replied:
Your clarify is clear, I have understood all your opinions, thanks.

cclerget replied:
Cool ! I also pushed changes related to https://github.com/opencontainers/runc/pull/3923#discussion_r1264319356

lifubang commented:
@cclerget I think this patch is helpful to solve the issue #3922 . 

I also think that the existence of [this patch](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76) has its meaning.
Maybe in most situation, people may need the behavior of this kernel's patch, but after your patch, people have no chance to let the kernel distribute tasks to different cpus, because your patch masks the kernel's patch.

cclerget commented:
@lifubang I understand your concerns, what if I restrict the scope of the patch by calling `getEligibleCPU` only when `kernel.sched_rt_runtime_us` has `-1` value ? When set to `-1`, it gives zero room for kernel to switch in kernel space on a CPU where a real-time application is running on (under scheduler real-time policies) and do not make any context switch.

This setting value is part of the problem that I'm attempting to address, when all tasks (except the slow-thread) are running under scheduler a real-time policy within a cgroup cpuset, the kernel patch will arbitrary pin `runc` process to one of the CPU during cgroup transition, if `runc` is not pinned to the CPU running the slow-thread (usually running on the first CPU of the cpuset for a deterministic placement), `runc` will get stuck and won't be able to enter in container with the side effect of locking down the system by acquiring a lock in cgroup subsystem during transition, the lock is not released until the `runc` process has been placed in scheduler CPU queue, which do not happen in this case until the real-time thread exit or make a context switch (not the case).

So it might make sense to also check this value and only call `getEligibleCPU` for this specific case. This should drastically reduce the scope of the patch, as many setup won't use this setting, and people using it are aware of consequences, so I think this patch will be required to restore the previous deterministic task placement of the kernel in such setup where this kernel patch is not appropriate, it is useful in most situations but not this one where random task placement has bad consequences

lifubang commented:
> what if I restrict the scope of the patch by calling `getEligibleCPU` only when `kernel.sched_rt_runtime_us` has `-1` value ?

I think it's a good idea.

cclerget commented:
Changes applied

lifubang commented:
> Changes applied

@cclerget Thanks. But sorry, you should rebase to resolve conflicts.

cclerget commented:
@lifubang done

lifubang commented on libcontainer/container_linux.go:
```diff
@@ -2358,3 +2378,120 @@ func requiresRootOrMappingTool(c *configs.Config) bool {
+//   - when there is not cpuset cores: no eligible CPU
+//   - when there is not isolated cores: no eligible CPU
+//   - when cpuset cores are all isolated cores: return the first CPU of the cpuset
+//   - when cpuset cores are mixed between housekeeping/isolated cores: return the
```
There is another situation: when all cpuset cores are not isolated cores.
This function should return -1, false, nil.

lifubang replied:
And add a test case to cover this situation, thanks.

cclerget replied:
This is covered by `isolatedCores > 0` https://github.com/opencontainers/runc/pull/3923/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2496 which is greater than 0 when there at least one core present in both sets.

And here the corresponding test case https://github.com/opencontainers/runc/pull/3923/files#diff-a632fc3b003cd903187e1091e7699141c62c87887c4a5e5ab17f99cb3de427c0R392

cclerget replied:
I updated the function comment as a covered case

lifubang replied:
To simplify this function, I have another idea. Could you consider break the for loop and return when isolatedCores>0 && affinityCore>-1?

lifubang replied:
> Yes, you covered it.

You have covered these two cases in one case:
1. cpuset: 0-1, isolated: 10-11
2. cpuset: 10-11, isolated: 0-1

cclerget replied:
> Yes, you covered it. My original thought is that we should set affinityCore to -1, if isolatedCores==0, or combine this two returns to one? Keep affinityCore, remove isolatedCores?

Fixed

> To simplify this function, I have another idea. Could you consider break the for loop and return when isolatedCores>0 && affinityCore>-1?

As we need to handle the mixed case, we can only return when mixed case is detected and isolatedCores > 0, it's covered here https://github.com/opencontainers/runc/pull/3923/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2483

Globally, the loop has been a bit refactored to be less verbose. Thanks !

cclerget replied:
> You have covered these two cases in one case:

    cpuset: 0-1, isolated: 10-11
    cpuset: 10-11, isolated: 0-1

Fixed too

cclerget commented:
@lifubang I realized that I forgot to remove the print line at https://github.com/opencontainers/runc/commit/a3c5576990cd33c7fc0fa6500f30c771af8ff501#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2386, I just removed it with this change https://github.com/opencontainers/runc/compare/a3c5576990cd33c7fc0fa6500f30c771af8ff501..00b21903d5c44c0b5220ca95375f675a2a08891a, sorry about that

lifubang commented:
Thanks, @kolyshkin PTAL

lifubang commented:
Need rebase @cclerget

lifubang commented:
@opencontainers/runc-maintainers PTAL
@cclerget has rebased many times. I think this patch is very important for some real-time nodes.

lifubang commented:
@cclerget I have no time to dive into the code design, do you know there is a way to move all your code from `c` to `golang`?

cclerget commented:
@lifubang Will look if I can move that into `setnsProcess.start`

lifubang commented on libcontainer/process_linux.go:
```diff
@@ -81,9 +84,40 @@ func (p *setnsProcess) signal(sig os.Signal) error {
+		}
+	}
+
+	err := startCommandWithCPUAffinity(p.cmd, cpuAffinity)
```
I think maybe we should move the `cpuAffinity` check to here to make the code more readable:
```
	var err Error
	if cpuAffinity < 0 {
		err = cmd.Start()
	} else {
		err = startCommandWithCPUAffinity(p.cmd, cpuAffinity)
        }
```

lifubang commented on libcontainer/process_linux.go:
```diff
@@ -81,9 +84,40 @@ func (p *setnsProcess) signal(sig os.Signal) error {
+		} else if cg.CpusetCpus != "" {
+			cpuAffinity, err = getCPUAffinity(os.DirFS("/"), cg.CpusetCpus)
+			if err != nil {
+				return fmt.Errorf("setting CPU affinity: %w", err)
```
setting -> getting

lifubang commented:
Thanks, overall still LGTM.
@cclerget Could you please add an integration test to read and check cpuAffinity value in the container with some mock data in `https://github.com/opencontainers/runc/blob/main/libcontainer/integration/exec_test.go`?
@AkihiroSuda Could you review this solution to check whether this is still LGTY or not before @cclerget squashing all the commits.

dqminh commented:
Can you explain more what does "temporary" mean here ? If we called `sched_setaffinity` manually to the first CPU i suppose we are working around the random distribution of tasks and the affinity to first CPU is permanent, am i understanding this correctly ?

cclerget commented:
@lifubang 
> Could you please add an integration test to read and check cpuAffinity value in the container with some mock data in https://github.com/opencontainers/runc/blob/main/libcontainer/integration/exec_test.go?

Now the question is, do you just want the case where there is no eligible CPU to ensure to not mess up with affinity for most case, or do you also want to test the case for an eligible CPU by mocking `/sys/devices/system/cpu/isolated` and `/proc/sys/kernel/sched_rt_runtime_us` ? For the latter case I could add an `fs.FS` to `libcontainer.Process`, if it sounds reasonable ?

@dqminh The call to `sched_setaffinity` just set the CPU affinity mask for a short period hence "temporary", when the runc process is placed in the container cgroup it is joining, cgroup set (or reset) the CPU affinity mask for the process to `cpuset.cpus` value, the process still run on the first CPU set until the kernel scheduler decide to move it to another CPU of `cpuset.cpus` list if necessary and possible, this restores the same behavior as before the kernel patch.

lifubang commented:
> For the latter case I could add an `fs.FS` to `libcontainer.Process`, if it sounds reasonable ?

I have a strange idea:
1. mount -o bind ./sched_rt_runtime_us /proc/sys/kernel/sched_rt_runtime_us
2. mount -o bind ./isolated /sys/devices/system/cpu/isolated
3. integration test
4. umount /proc/sys/kernel/sched_rt_runtime_us
5. umount /sys/devices/system/cpu/isolated

cclerget commented:
@lifubang the hacky way yeah, that should do it, hoping to not interfere with other tests running in parallel in the future, there is few chance but we never know

kolyshkin commented on libcontainer/message_linux.go:
```diff
@@ -22,6 +22,10 @@ const (
+	MountSourcesAttr uint16 = 27290
+	IdmapSourcesAttr uint16 = 27291
+	TimeOffsetsAttr  uint16 = 27292
+	CPUAffinityAttr  uint16 = 27293
```
Why is this here?

kolyshkin replied:
I mean, `MountSources` and `IdmapSources`.

kolyshkin commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -110,6 +113,10 @@ struct nlconfig_t {
 #define TIMENSOFFSET_ATTR	27290
+#define MOUNT_SOURCES_ATTR	27290
+#define IDMAP_SOURCES_ATTR	27291
+#define TIMENSOFFSET_ATTR	27292
```
ditto

cclerget replied:
same

kolyshkin commented on docs/isolated-cpu-affinity-transition.md:
```diff
@@ -0,0 +1,74 @@
+__WARNING:__ `definitive` requires a kernel >= 6.2, also works with RHEL 9 and
+above.
+
+### How it works ?
```
nit: extra space

cclerget replied:
I don't get it, could you suggest the patch please ?

kolyshkin reviewed (commented):
One general comment: please ensure the comments you're adding start with an Uppercased letter and end with a period.

Otherwise it's harder to maintain long-term.

kolyshkin commented on tests/integration/exec.bats:
```diff
@@ -340,3 +340,143 @@ EOF
+	mount --bind "$tmp/cmdline" /proc/cmdline
+
+	runc exec test_isolated_bad_transition true
+	runc_status=$status
```
You don't have to save `$status`, as `umount` command won't change it.

cclerget replied:
fixed

kolyshkin commented on libcontainer/system/kernelparam/lookup.go:
```diff
@@ -0,0 +1,52 @@
+	"unicode"
+)
+
+type KernelBootParam string
+
+const (
+	IsolatedCPUAffinityTransition KernelBootParam = "runc.exec.isolated-cpu-affinity-transition"
+	NohzFull                      KernelBootParam = "nohz_full"
```
I'd just use the `string` directly, and remove these consts (move them to code that's using `LookupKernelBootParameters`)

cclerget replied:
fixed

kolyshkin commented on libcontainer/system/kernelparam/lookup.go:
```diff
@@ -0,0 +1,52 @@
+import (
+	"io/fs"
+	"strings"
+	"unicode"
```
I _think_ that kernel boot parameters is guaranteed to be in ASCII and not UTF-8, so maybe no `unicode` usage is required. Have to check that though.

cclerget replied:
fixed

cclerget commented:
> One other observation -- you add some code to the first commit, when remove most of it in the second one. Perhaps you can rework this to avoid doing that, as otherwise it's hard to review commit by commit.

Sorry about that, so I squashed the two commits and I'm pushing separate commits for each review rounds.

An additional note, my PR includes a similar fix to https://github.com/opencontainers/runc/pull/4041 which should work for all kernel, I don't know when the patch referred [here](https://lore.kernel.org/lkml/20220922180041.1768141-1-longman@redhat.com/) to reset the CPU affinity is going to be merged in kernel upstream, but even if it is, it won't fill the gap for a range of kernel version since 6.2

cclerget commented:
Hello @kolyshkin @lifubang @AkihiroSuda , any chance to get it reviewed please ?

arun2arunraj commented:
@AkihiroSuda : Any idea when this PR will be merged and released for public usage ?

AkihiroSuda commented:
This PR is still marked as a draft, so this is not considered ready to be merged

AkihiroSuda commented on docs/isolated-cpu-affinity-transition.md:
```diff
@@ -0,0 +1,74 @@
+## Isolated CPU affinity transition
+
+The introduction of the kernel commit 46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76
```
Version should be shown here

cclerget replied:
done

AkihiroSuda commented on docs/isolated-cpu-affinity-transition.md:
```diff
@@ -0,0 +1,74 @@
+Runc introduces a way to restore this behavior via a dedicated kernel boot
+parameter:
+
+`runc.exec.isolated-cpu-affinity-transition`
```
Why not use:
- an annotation in config.json
- a CLI flag for runc
- env var

cclerget replied:
This PR essentially targets kubernetes environments using realtime tuned profiles, that's why I choose a kernel param as it could be added as part of the tuned profile. I thought about annotations but I'm a bit hesitant for two reasons:
- for security reasons, in k8s environments with untrusted users, giving the possibility to alter such behavior through pod annotations might not be ideal
- interactions between annotations and runtimes in k8s CRI as highlighted [here](https://github.com/kubernetes/cri-api/blob/30022f215cd7c176a48d567186d153476fc46f7b/pkg/apis/runtime/v1/api.proto#L462-L469).

Regarding CLI flag and environment variable, it would require additional configuration by using a dedicated wrapper to pass the CLI flag or modifying systemd unit for the environment variable, plus the environment variable could not be passed correctly to runtime by CRI or others, those options are not really convenient IMHO.

I found kernel param approach to be the less intrusive and less coupled

AkihiroSuda replied:
I mentioned an annotation in config.json, not an annotation in Kubernetes pod
https://github.com/opencontainers/runtime-spec/blob/main/config.md#annotations

AkihiroSuda replied:
Also curious why not just add a new proper JSON field to config.json ?

cclerget replied:
Yes I understood, this fix needs to be enabled either system wide or per container primary in kubernetes environment running realtime applications without dealing with the OCI runtime spec, basically it should be deployed in environments with current tools like containerd/kubelet without any modifications at this level, so modifying the spec is not really an option. I mentionned about pod annotations as those can be passed to runtimes by CRI implementations like containerd/cri-o via OCI runtime spec annotations

AkihiroSuda replied:
Pod annotations are not OCI annotations

cclerget replied:
Right, I'm just talking about how it would work from a kubernetes pov to enable this feature by using an OCI annotation like you've suggested instead of using the kernel param.

AFAIK containerd allow to pass (un)filtered pod annotations for CRI container runtime via `pod_annotations`:

```
[plugins."io.containerd.grpc.v1.cri".containerd]
  default_runtime_name = "runc"
  [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]
    [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
      runtime_type = "io.containerd.runc.v2"
      base_runtime_spec = "/etc/containerd/cri-base.json"
      pod_annotations = ["*", "*/*"]
      container_annotations = ["*", "*/*"]
```

While looks like cri-o automatically pass pod annotations and store them as OCI annotations.

Such that a pod annotated with `org.opencontainers.runc.exec.isolated-cpu-affinity-transition` by example would enable the feature for runc for exec probes or `kubectl exec`.

Is it this approach that you would prefer ?

cclerget replied:
@AkihiroSuda ^^

AkihiroSuda replied:
> While looks like cri-o automatically pass pod annotations and store them as OCI annotations.

If this is true, it is a serious bug of cri-o.

> `org.opencontainers.runc.exec.isolated-cpu-affinity-transition`

SGTM

haircommander replied:
how else would a k8s user specify annotations for the oci runtime?

AkihiroSuda replied:
> how else would a k8s user specify annotations for the oci runtime? just image annotations??

No, non-admin user mustn't be able to specify OCI annotations that change the behavior of the runtime.

cclerget replied:
@AkihiroSuda changes committed in https://github.com/opencontainers/runc/pull/3923/commits/197b60daec3f9257811f350da46f15adec24a8d4 to use OCI annotation instead of kernel param

lifubang commented:
> One other observation -- you add some code to the first commit, when remove most of it in the second one. Perhaps you can rework this to avoid doing that, as otherwise it's hard to review commit by commit.

@kolyshkin @AkihiroSuda @cclerget 
Sorry for delay reply, this is becuase https://github.com/opencontainers/runc/pull/3923#issuecomment-1709337514 . @cclerget wanted us to choose which one is better? C or Go?

cclerget commented:
Hello @kolyshkin @lifubang @AkihiroSuda, I marked the PR as ready for review, the code is actually
well covered by various tests, the feature is disabled by default and can be enabled only with this
new kernel param such that it doesn't impact the current behavior.

I would also like to reiterate, there is also another important fix similar to https://github.com/opencontainers/runc/pull/4041 but covering all kernel since kernel 6.2. To give you an idea of the issue, you can add this [test](https://github.com/opencontainers/runc/pull/3923/files#diff-b03819d46bd922c514657ed0e4b8be870459566df28338bf30f6d8fcbd8f9719R448) to the main branch and run the tests on RHEL 9 or on a system with a kernel >= 6.2 and see the container process assigned to the first CPU only instead of all when runc is executed under specific CPU affinity.

I hope we could move forward quickly on this PR once you are all agree.

AkihiroSuda reviewed (changes requested):
I'm still not convinced why we have to introduce a kernel cmdline specifically for this

AkihiroSuda commented on docs/isolated-cpu-affinity-transition.md:
```diff
@@ -0,0 +1,148 @@
+Runc introduces a way to restore this behavior by adding the following
+annotation to the container runtime spec (`config.json`):
+
+`org.opencontainers.runc.exec.isolated-cpu-affinity-transition`
```
This list should be updated too
https://github.com/opencontainers/runc/blob/8e69225e2e5f25c46e79cf49935db362ead8f496/features.go#L67-L71

cclerget replied:
Ok, do you know if CRI implementations like containerd are using this list actually to automatically exclude annotations ?

AkihiroSuda replied:
Not yet

AkihiroSuda commented on docs/isolated-cpu-affinity-transition.md:
```diff
@@ -0,0 +1,148 @@
+metadata:
+  name: demo-pod
+  annotations:
+    org.opencontainers.runc.exec.isolated-cpu-affinity-transition: "temporary"
```
This is not needed for containerd IIUC

cclerget replied:
I guess it's related to your comment about `container_annotations`, because I don't see how runc could enable the feature without that

AkihiroSuda commented on docs/isolated-cpu-affinity-transition.md:
```diff
@@ -0,0 +1,148 @@
+    [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
+      runtime_type = "io.containerd.runc.v2"
+      base_runtime_spec = "/etc/containerd/cri-base.json"
+      pod_annotations = ["org.opencontainers.runc.exec.isolated-cpu-affinity-transition"]
```
`container_annotations` might be better here, as the annotation does not really relate to Kubernetes pods
https://github.com/containerd/containerd/blob/v1.7.14/docs/cri/config.md?plain=1#L310

cclerget replied:
From my tests only `pod_annotations` is working and pass the annotation to sandbox and container, `container_annotations` are coming from device plugins and kubernetes internals, like stated in the comment this is not controlled by users, preventing users to enable the feature on a per-pod basis

AkihiroSuda commented on docs/isolated-cpu-affinity-transition.md:
````diff
@@ -0,0 +1,148 @@
+    image: registry.com/demo:latest
+```
+
+#### Deployment template example
````
This might be redundant

cclerget replied:
I will remove it

AkihiroSuda commented:
Please squash the commits when this PR is ready to be merged

cclerget commented:
Ok I've pushed the requested changes, I haven't applied changes related to your comments https://github.com/opencontainers/runc/pull/3923#pullrequestreview-1962335750 and https://github.com/opencontainers/runc/pull/3923#pullrequestreview-1962330692, please let me know if you disagree with my comments, otherwise I'll go ahead and squash commits

cclerget commented:
@AkihiroSuda commits squashed

kolyshkin commented on libcontainer/cgroups/fs/fs.go:
```diff
@@ -263,3 +265,26 @@ func (m *Manager) OOMKillCount() (uint64, error) {
+	}
+
+	// Iterates until it goes to the cgroup root path.
+	for path := filepath.Clean(cpusetPath); path != defaultCgroupRoot; path = filepath.Dir(path) {
```
AFAIK filepath.Clean is not needed here because m.Path returns cleaned path.

kolyshkin commented on libcontainer/cgroups/fs/fs.go:
```diff
@@ -263,3 +265,26 @@ func (m *Manager) OOMKillCount() (uint64, error) {
+		return ""
+	}
+
+	// Iterates until it goes to the cgroup root path.
```
Maybe makes sense to add something like "needed for containers in which cpuset controller is not enabled -- in this case a parent cgroup is used" -- if my understanding is correct.

kolyshkin commented on libcontainer/process_linux.go:
```diff
@@ -133,8 +137,58 @@ func (p *setnsProcess) start() (retErr error) {
+				annotations,
+			)
+			if err != nil {
+				// Close the pipe to not be blocked in the parent.
+				p.comm.closeChild()
```
ditto

cclerget replied:
same

kolyshkin commented on libcontainer/process_linux.go:
```diff
@@ -995,5 +1069,196 @@ func setIOPriority(ioprio *configs.IOPriority) error {
+	// Use a goroutine to dedicate an OS thread.
+	go func() {
+		cpuSet := new(unix.CPUSet)
+		cpuSet.Zero()
```
AFAIK this is not needed, in Go everything is initialized to default values (0s in this case), and you've just instantiated a new CPUSet.

kolyshkin commented on tests/integration/exec.bats:
```diff
@@ -340,3 +340,168 @@ EOF
 }
+
+@test "runc exec with isolated cpus affinity temporary transition [cgroup cpuset]" {
+	requires root
```
need to add `cgroups_cpuset` to the requires list.

kolyshkin replied:
Same for other tests

kolyshkin commented on tests/integration/exec.bats:
```diff
@@ -340,3 +340,168 @@ EOF
+	# set all online cpus as isolated
+	echo "nohz_full=$all_cpus" >"$tmp/cmdline"
+
+	mount --bind "$tmp/cmdline" /proc/cmdline
```
who's undoing this mount?

kolyshkin replied:
my bad, missed the line below.

kolyshkin commented on tests/integration/exec.bats:
```diff
@@ -340,3 +340,168 @@ EOF
+
+	tmp=$(mktemp -d "$BATS_RUN_TMPDIR/runc.XXXXXX")
+
+	local all_cpus
+	all_cpus="$(cat /sys/devices/system/cpu/online)"
+
+	update_config ".linux.resources.cpu.cpus = \"$all_cpus\""
+
+	# set temporary isolated CPU affinity transition
+	update_config '.annotations += {"org.opencontainers.runc.exec.isolated-cpu-affinity-transition": "temporary"}'
+
+	local mems
+	mems="$(cat /sys/devices/system/node/online 2>/dev/null || true)"
+	[[ -n $mems ]] && update_config ".linux.resources.cpu.mems = \"$mems\""
```
Perhaps you can separate this into a function (at least the mems and all_cpus part).

kolyshkin commented on tests/integration/exec.bats:
```diff
@@ -340,3 +340,168 @@ EOF
+
+	load /etc/os-release
+
+	# fix unbound variable in condition below
+	PLATFORM_ID=${PLATFORM_ID:-}
```
nit: can you use `VERSION_ID` instead, it looks easier?

cclerget commented on libcontainer/process_linux.go:
```diff
@@ -133,8 +137,58 @@ func (p *setnsProcess) start() (retErr error) {
+	if len(p.manager.GetPaths()) > 0 {
+		// Get the target container cgroup.
+		if cg, err := p.manager.GetCgroups(); err != nil {
+			// Close the pipe to not be blocked in the parent.
+			p.comm.closeChild()
```
Not enough, the defer closes the parent side of the pipe, we also need to close the child side otherwise the process get stuck

cclerget commented:
@kolyshkin addressed your comments in https://github.com/opencontainers/runc/pull/3923/commits/ea7b6c03c026310013f6cc16c8a03b1cee3fcfce, will squash the commits after approval

andreaskaris commented on docs/isolated-cpu-affinity-transition.md:
```diff
@@ -0,0 +1,126 @@
+The introduction of the kernel commit 46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76
+in 5.7 has affected a deterministic scheduling behavior by distributing tasks
+across CPU cores within a cgroups cpuset. It means that some runc operations
+like `runc exec` might be impacted under some circumstances, by example when
```
Not a review, just a question (but maybe you'll decide to clarify this further). Without looking at the code, and only at this piece of documentation (and at the commit message):
* will this only improve the behavior of threads launched with `runc exec`?  (if so, then the documentation should not be "some runc operations like")
* otherwise, what other commands or situations will benefit from this patch? (i.e., in which cases do I want to annotate my pods with this annotation to revert to the behavior pre 46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76)
* related to the above, will this change the behavior of any new process spawned in the container (an example would be a DPDK application using the vhost device to create kernel vhost threads, would those still float around freely, or be sent to the first CPU with the annotation in place

andreaskaris replied:
Awesome, thanks for the clarification and thanks for your work on this!

cclerget commented:
@kolyshkin changes ok with you ?

kolyshkin reviewed (approved):
LGTM; please squash the commits

cclerget commented:
Done thanks !

MatthewHink commented:
Thank goodness this is finally merged! Really appreciate your help everyone!

kolyshkin commented:
The more I look into this the more I think I did a bad job reviewing this, and it needs to be redone in a different way:
1. There's too much logic here figuring out which CPUs to use. Runc is a low level tool and is not supposed to be that "smart".
2. What's worse, this logic is executed on every exec, making it slower.
3. Some of the logic in `(*setnsProcess).start` is executed even if no annotation is set, thus making ALL execs slow.
4. As pointed out in https://github.com/opencontainers/runtime-spec/pull/1252, if we want to support this across different runtimes (e.g. crun and runc), this should not be an annotation, but rather a process parameter.

Let's fix this:
1. Revert this PR.
2. Move some of the functionality (determining which CPU to pin exec to) to upper level runtime (cri-o/containerd).
3. Open a PR in runtime-spec proposing the changes to runc/crun.
4. Open a PR to runc implementing that proposal.

NeilHanlon commented:
while I recognize that this is a complex issue, I am unsure if reverting this is the best path forward, considering implementations of this are already deployed and in use.

This review has taken an exceedingly long time and we are now on the cusp of even *more* review.

It would be disappointing if we have to start this whole process over again.

kolyshkin commented:
> while I recognize that this is a complex issue, I am unsure if reverting this is the best path forward, considering implementations of this are already deployed and in use.
> 
> This review has taken an exceedingly long time and we are now on the cusp of even _more_ review.
> 
> It would be disappointing if we have to start this whole process over again.

I have to admit I did a sloppy job reviewing this; yet this is not in any of the released runc versions (and this is why it needs to be reverted now, before we officially release it).

Now, this has to be re-implemented in the right way, starting from runtime-spec (see https://github.com/opencontainers/runtime-spec/pull/1253), then in runtimes (such as cri-o and containerd), when in low level runtimes (such as runc and crun).

Any help (esp in cri-o and containerd) is appreciated.

kolyshkin commented:
@cclerget I am sorry this had to be reverted, but now when  opencontainers/runtime-spec#1253 is merged we can work on reviving it if you're still interested. Now, 90% of the work is to be done by containerd/cri-o (they should do the work of converting pod annotation (what used to be `org.opencontainers.runc.exec.isolated-cpu-affinity-transition`) to `process.ExecCPUAffinity` setting for runc, and runc should only enforce the affinity.

Let me know if you still want to work on this and we'll figure it out.

cclerget commented:
@kolyshkin It's unfortunate but understand the rational behind this decision.

I'm more concerned about the revert of the fix not directly related to the original issue like mentioned in my comment https://github.com/opencontainers/runc/pull/3923#issuecomment-1948291803.

It has also been noted by the customer when using a tuned profile like CPU-partitioning, the CPU affinity patch at least on kernel < 6.2 or RHEL < 9 doesn't fix the runc "noisy neighbor" behavior after cgroup transition, on those kernels and systems `execCPUAffinity.final` won't be effective without a rework of cgroup handling, I'm not sure if this would work with the various cgroup managers, but the idea is to create a child cgroup (let's call it `nsexec`) under the container cgroup when `execCPUAffinity.initial` is set and `execCPUAffinity.final` is set and different, in order to put the runc exec process in this `nsexec` child cgroup, such that this child cgroup could set a more restricted cpuset reflecting `execCPUAffinity.final` when different from `execCPUAffinity.initial` and restrict the executed process to this specific cpuset to remediate to the "noisy neighbor" behavior. Does that make sense ?

I'd like to but right now I haven't much room to work on it, maybe in two weeks

jsco2t commented:
@kolyshkin has there been any commitment by `containerd/cri-o` as to when they might land their portion of this solution? I looked through the recent issues in `containerd` but didn't see anything to suggest to me that they were doing planning around this feature area (though I certainly could have missed it).
